packages/clay-core/src/tree-view/useItem.tsx

```diff
@@ -0,0 +1,88 @@
+/**
+ * SPDX-FileCopyrightText: © 2021 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import React, {useContext, useEffect, useRef} from 'react';
+import {useDrag, useDrop} from 'react-dnd';
+
+import {useTreeViewContext} from './context';
+
+type Value = {
+	[propName: string]: any;
+	indexes: Array<number>;
+	key: React.Key;
+};
+
+type Props = {
+	children: React.ReactNode;
+	value: Omit<Value, 'indexes'> & {index: number};
+};
+
+const ItemContext = React.createContext<Value>({} as Value);
+
+function getKey(key: React.Key) {
+	return `${key}`.replace('.$', '');
+}
+
+export function ItemContextProvider({children, value}: Props) {
+	const {items, open, reorder, selection} = useTreeViewContext();
+	const {indexes: parentIndexes = [], key: parentKey} = useItem();
+
+	const keyRef = useRef(getKey(value.key));
+
+	const childRef = useRef(null);
+
+	useEffect(() => selection.mount(keyRef.current, parentKey), [
+		selection.mount,
+		keyRef,
+		parentKey,
+	]);
+
+	const item: Value = {
+		...value,
+		indexes: [...parentIndexes, value.index],
+		key: keyRef.current,
+	};
+
+	const [, drag] = useDrag({
+		item: {
+			...item,
+			type: 'treeViewItem',
+		},
+	});
+
+	const [, drop] = useDrop({
+		accept: 'treeViewItem',
+		collect: (monitor) => ({
+			canDrop: monitor.canDrop(),
+			overTarget: monitor.isOver({shallow: true}),
+		}),
+		drop(dragItem: unknown, monitor) {
+			if (monitor.didDrop()) {
+				return;
+			}
+
+			reorder((dragItem as Value).indexes, item.indexes);
```

Comment:
Yes, that's a good idea. I personally think that if we're able to determine at which position the item should be added it would be ideal, but the basic idea was that you could reorder nodes in the tree, I'll also check how far we can improve this.

Reply:
Hey @matuzalemsteles, I've added https://github.com/liferay/clay/pull/4255/commits/eacefd2ec5d633938a21fa7d12c67024d9e0d6e5 which makes it possible to "reorder" child nodes in the same "root/parent" node. Might need some cleaning up but I still wanted you to take a quick look

Reply:
Oh perfect, I'm going to take a look at this commit.